src/login.ts

```diff
@@ -0,0 +1,24 @@
+import { compare } from 'bcrypt';
+import { appDataSource } from './data-source';
+import { User } from './entity/user';
+import { LoginInput } from './schema';
+import { BaseError } from './base-error';
+
+export async function getUserFromLoginInput(credentials: LoginInput) {
+  const userRepository = appDataSource.getRepository(User);
+  const user = await userRepository.findOne({ where: { email: credentials.email } });
+
+  if (user) {
+    if (await compare(credentials.password, user.password)) {
+      return user;
+    } else {
+      throw new BaseError('Invalid credentials', 400, 'Wrong password');
+    }
+  } else {
+    throw new BaseError('Invalid credentials', 400, 'No user with this email');
+  }
```

Comment:
Aqui eu vejo 2 abordagens:

1. Seu sistema vai informar que o e-mail não existe na base (pela mensagem ele parece já fazer isso): seria melhor retornar 404, de not-found
2. Seu sistema vai esconder se aquele e-mail existe ou não na base: retorna 401 com a mesma mensagem do erro anterior

Eu particularmente acho que o botinho meio que influencia vcs a irem pela abordagem 1. A maioria dos nosso projetos acabam indo por ela, se não me engano. Alguns outros que são mais sistemas que tem menos utilização marketeira vão pela 2, pela maior segurança.

O que acha?

Reply:
Optei por (1) com a mensagem "Wrong email or password"